### PR TITLE
Dashboards: Add v1alpha1 and v2alpha1 conversion

### PR DESCRIPTION
**Pull request by toddtreece**
**What is this feature?**

This PR adds in dashboard v1 and v2, that are essentially the same as v0, in order to give something to work upon. It also stubs out a [conversion function](https://github.com/grafana/grafana/pull/96415/files#diff-6fe4c76871fdd8371bd590b3f2275972091d0a48055991912f41b8e98fdc8166R169) between the versions

example: /apis/dashboard.grafana.app/v0alpha1/namespaces/default/dashboards/fe3yf9awmdkowe/dto
Screenshot 2024-11-14 at 4 23 38 PM

example: /apis/dashboard.grafana.app/v1alpha1/namespaces/default/dashboards/fe3yf9awmdkowe/dto
Screenshot 2024-11-14 at 4 23 54 PM

example: /apis/dashboard.grafana.app/v2alpha1/namespaces/default/dashboards/fe3yf9awmdkowe/dto
Screenshot 2024-11-14 at 4 24 08 PM

### CODE REVIEW

File: pkg/apimachinery/apis/common/v0alpha1/unstructured.go [hunk 1]
```diff
@@ -14,7 +14,7 @@ import (
 type Unstructured struct {
 	// Object is a JSON compatible map with string, float, int, bool, []interface{},
 	// or map[string]interface{} children.
-	Object map[string]any
+	Object map[string]interface{}
```
Comment (toddtreece) [inline pkg/apimachinery/apis/common/v0alpha1/unstructured.go:17]: `kube-openapi` generation will fail with this error when `map[string]any` is used, but is fine with `map[string]interface{}` 🙃 

```
Generating openapi code for 1 targets
E1114 08:51:34.777207 1599037 openapi.go:331] Error when generating: Object, Object map[string]any
2024/11/14 08:51:34 OpenAPI code generation error: failed executing generator: some targets had errors: failed to generate map property in github.com/grafana/grafana/pkg/apimachinery/apis/common/v0alpha1.Unstructured: Object: map Element kind Interface is not supported in map[string]any
```

File: pkg/apis/dashboard/v1alpha1/conversion.go [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+	common "github.com/grafana/grafana/pkg/apimachinery/apis/common/v0alpha1"
+)
+
+func Convert_v0alpha1_Unstructured_To_v1alpha1_DashboardSpec(in *common.Unstructured, out *DashboardSpec, s conversion.Scope) error {
+	out.Unstructured = *in
+
+	t := in.Object["title"]
+	title, ok := t.(string)
+	if !ok {
+		klog.V(5).Infof("dashboard v2alpha1 title field is not a string %s", t)
```
Comment (charandas) [inline pkg/apis/dashboard/v1alpha1/conversion.go:16]: This log line makes me think its not a failed assertion, but I think its trying to call out a failed assertion based on my inspections of types.go, where it is indeed a `string`. What am I missing?

**Review by charandas — APPROVED** [review]
The boilerplate looks good to me here!

File: pkg/apis/dashboard/v1alpha1/conversion.go (hunk 2 above)
Comment (toddtreece) [reply pkg/apis/dashboard/v1alpha1/conversion.go]: updated with [5133796](https://github.com/grafana/grafana/pull/96415/commits/51337969156aaa2c199e95199a6ffe8643ff14b9). thoughts?

**Review by ryantxu — APPROVED** [review]
This works great -- amazing.

I am very nervous about adding v2 schema into the current dashboard SQL table -- that will be pretty bonkers with totally weird failure modes.

Perhaps something like https://github.com/grafana/grafana/pull/96586 -- but that is not really workign as expected yet 😢 

This is a big step forward, so lets merge and we can make sure writing v2 is only super dev machines only

**Pull request merged**